malariagen_data/anoph/hap_freq.py

```diff
@@ -0,0 +1,391 @@
+            cohort_key_str = cohort.taxon + "_" + cohort.area + "_" + str(cohort.period)
+            # We reset all frequencies, counts to 0 for each cohort, nobs is set to the number of haplotypes
+            n_samples = cohort.size
+            hap_freq = {k: 0 for k in f_all.keys()}
+            hap_count = {k: 0 for k in f_all.keys()}
+            hap_nob = {k: 2 * n_samples for k in f_all.keys()}
+            assert n_samples >= min_cohort_size
+            sample_indices = group_samples_by_cohort.indices[cohort_key]
+            loc_coh = [i in sample_indices for i in range(0, gt.shape[1])]
+            gt_coh = allel.GenotypeDaskArray(da.compress(loc_coh, gt, axis=1))
+            gt_hap = gt_coh.to_haplotypes().compute()
+            f, c, o = haplotype_frequencies(gt_hap)
+            # The frequencies and counts of the observed haplotypes are then updated, so are the nobs but the values should actually stay the same
+            hap_freq.update(f)
+            hap_count.update(c)
+            hap_nob.update(o)
+            count_cols["count_" + cohort_key_str] = list(hap_count.values())
+            freq_cols["frq_" + cohort_key_str] = list(hap_freq.values())
+            nobs_cols["nobs_" + cohort_key_str] = list(hap_nob.values())
+
+        df_freqs = pd.DataFrame(freq_cols, index=hap_freq.keys())
+
+        # Compute max_af.
+        df_max_af = pd.DataFrame({"max_af": df_freqs.max(axis=1)})
+
+        df_counts = pd.DataFrame(count_cols, index=hap_count.keys())
+
+        df_nobs = pd.DataFrame(nobs_cols, index=hap_nob.keys())
+
+        # Build the final dataframe.
+        df_haps = pd.concat([df_freqs, df_counts, df_nobs, df_max_af], axis=1)
+
+        df_haps_sorted = df_haps.sort_values(by=["max_af"], ascending=False)
+        df_haps_sorted["label"] = ["H" + str(i) for i in range(len(df_haps_sorted))]
+
+        # Reset index after filtering.
+        df_haps_sorted.set_index(keys="label", drop=True)
+
+        # Build the output dataset.
+        ds_out = xr.Dataset()
+
+        # Cohort variables.
+        for coh_col in df_cohorts.columns:
+            ds_out[f"cohort_{coh_col}"] = "cohorts", df_cohorts[coh_col]
+
+        # Label the haplotypes
+        ds_out["variant_label"] = "variants", df_haps_sorted["label"]
+        # Event variables.
+        ds_out["event_frequency"] = (
+            ("variants", "cohorts"),
+            df_haps_sorted.to_numpy()[:, : len(df_cohorts)],
+        )
+        ds_out["event_count"] = (
+            ("variants", "cohorts"),
+            df_haps_sorted.to_numpy()[:, len(df_cohorts) : 2 * len(df_cohorts)],
+        )
+        ds_out["event_nobs"] = (
+            ("variants", "cohorts"),
+            df_haps_sorted.to_numpy()[:, 2 * len(df_cohorts) : -2],
+        )
+
+        # Add confidence intervals.
+        _add_frequency_ci(ds=ds_out, ci_method=ci_method)
+
+        return ds_out
+
+
+def _prep_samples_for_cohort_grouping(*, df_samples, area_by, period_by):
```

Comment:
This and all other functions below in the module are duplicates of functions already in the `snp_frq` module. Suggest to import from there rather than copying here.

Reply:
Thank @alimanfoo. Given that these functions are not haplotype or SNP specific, wouldn't it make more sense to move them to `util` as well?

Reply:
Yep, seems reasonable to me.